### DISCUSSION TITLE
Is it possible to reuse my JavaScript code into iOS

### DISCUSSION THREAD

**Discussion by Josema** [google/filament#7243] [General]
Hi, just wondering if is possible to reuse the [javascript triangle example](https://google.github.io/filament/webgl/tutorial_triangle.html) code to make it work in iOS or Android? (write once run everywhere).

If so, wouldn't be possible to make a port of Three.js or Babylon.js to make it work on iOS or Android rendered by filament? 

Maybe I'm a bit lost and I don't understand the scope of this project.

**Comment by poweifeng** [comment]
Not completely sure what the question means, but
 - You can certainly run web apps on Android and iOS by using WebView (or the iOS equivalent).  
 - Performance is the main difference between running an app in webview vs. "native".
 - AFAIK three.js and babylon.js are javascript 3D rendering libraries based on webgl
 - Filament is a rendering lib written in c++, and compiled for multiple platform (Android, iOS, web, desktop platforms like Windows, linux, macos). 
   - It's intended to be a PBR library for *mobile* platforms (iOS, Android)
   - Filament is compiled for the web through emscripten/wasm and porting from gles to webgl.

**Reply by Josema** [reply]
Yeah, sorry, I didn't explain myself correctly.

My question was about reusing code. I mean, imagine I want to create a simple video game using Filament. I would develop it first for the web because I'm familiar with JavaScript. So I was wondering if the same JavaScript code can be run on iOS and Android without too much extra effort.

This is essentially what Unity or Unreal Engine does. You only need to write your code once, and then you can export it easily to every platform.

**Reply by poweifeng** [reply]
For the case you described, you can try webview, but again, performance is probably not going to be great. 

For writing it once and porting it multiple platform, that's possible if your "once" code is in c++, but you'll have to go through some hoops to get port it to the web (i.e. wasm).

**Reply by Josema** [reply]
And how do you convert your c++ code to Java/Swift/Javascript. I mean, what is the process of porting? I'm just curious.

**Reply by poweifeng** [reply]
For Android, you can write native apps in c++ using the ndk.  
For iOS, ObjC is compatible with c++. 
For web,  you can leverage emscripten to compile c++ into wasm .
